ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ResponseOutputStream.java

```diff
@@ -55,13 +59,29 @@ public void write(final @NotNull byte[] bytes, final int offset, final int lengt
     final byte[] bytesToWrite = new byte[length];
     System.arraycopy(bytes, offset, bytesToWrite, 0, length);
     final Buffer buffer = Buffer.buffer(bytesToWrite);
+    blockIfWriteQueueFull();
     response.write(buffer);
   }
 
   @Override
   public void close() {
     response.end();
   }
+
+  private void blockIfWriteQueueFull() {
+    VertxUtils.checkIsWorker();
+    if (response.writeQueueFull()) {
+      final CompletableFuture<Void> cf = new CompletableFuture<>();
+      response.drainHandler(v -> cf.complete(null));
+      try {
+        cf.get(60, TimeUnit.SECONDS);
+      } catch (Exception e) {
```

Comment:
Should we check explicitly for `TimeoutException`? Since we don't expect `ExecutionException` to be thrown here.

Reply:
We could, but I don't think it adds much as we want to rethrow on any exception we receive.